### PR TITLE
add version api and show server version in cli

### PR DESCRIPTION
**Pull request by mxyng**
some minor refactor of the cmd package

Example: server and client are the same version
```
$ ollama --version
Your ollama version 0.0.0
```

Example: server and client have different versions
```
$ ollama --version
Your ollama version 99.99.99999
Warning: Your client version is 0.0.0
```

Example: server is not accessible
```
$ ollama --version
Warning: Your server is not accessible
Your client version is 0.0.0
```

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -914,8 +908,25 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
...
 
+func versionHandler(_ *cobra.Command, _ []string) {
+	fmt.Printf("ollama version %s\n", version.Version)
+
+	client, err := api.ClientFromEnvironment()
+	if err != nil {
+		return
+	}
+
+	serverVersion, err := client.Version(context.TODO())
```
Comment (BruceMacD) [inline cmd/cmd.go:919]: You could use the cmd.Context here too by updating the arguments

**Review by BruceMacD — COMMENTED** [review]
look good, one small comment

File: cmd/cmd.go (hunk 1 above)
Comment (mxyng) [reply cmd/cmd.go]: Good catch. I did this making the other context changes so must have missed it

File: cmd/cmd.go [hunk 2]
```diff
@@ -914,8 +908,25 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
...
+	if err != nil {
+		return
+	}
+
+	serverVersion, err := client.Version(cmd.Context())
+	if err != nil {
+		return
+	}
+
+	fmt.Printf("server version %s\n", serverVersion)
```
Comment (jmorganca) [inline cmd/cmd.go:924]: Could we only print this if a remote host is specified with `OLLAMA_HOST`? and then something like:

```suggestion
        hostVersion, err := client.Version(cmd.Context())
	if err != nil {
		return
	}

	fmt.Printf("host version %s\n", serverVersion)
```

**Comment by jmorganca** [comment]
small comment on wording/displaying the versions @pdevine will have good input on this one too

File: cmd/cmd.go (hunk 2 above)
Comment (jmorganca) [inline cmd/cmd.go:924]: Let’s print this only if they defer, and show “host version” since the user is presumably using OLLAMA_HOST here.

Comment (mxyng) [reply cmd/cmd.go]: I went with only printing this when the server version differs from the client. If your assertion is correct, the result is the same.

File: cmd/cmd.go [hunk 3]
```diff
@@ -919,27 +932,40 @@ func NewCLI() *cobra.Command {
 		CompletionOptions: cobra.CompletionOptions{
 			DisableDefaultCmd: true,
 		},
-		Version: version.Version,
+		PersistentPreRunE: func(cmd *cobra.Command, _ []string) error {
+			if !slices.Contains([]string{"serve", "version"}, cmd.Name()) {
+				return checkServerHeartbeat(cmd, nil)
```
Comment (pdevine) [inline cmd/cmd.go:937]: Abstracting this into the root command makes it more difficult for someone to reason about what's happening in each individual command, and it really only saves a few lines of code.

File: cmd/cmd.go [hunk 4]
```diff
@@ -908,8 +902,27 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
 	return nil
 }
 
+func versionHandler(cmd *cobra.Command, _ []string) {
+	fmt.Printf("ollama version %s\n", version.Version)
```
Comment (pdevine) [inline cmd/cmd.go:906]: This should be the host version and not the client version although we should still call it the "Ollama version" since it's the main thing.

File: cmd/cmd.go [hunk 5]
```diff
@@ -908,8 +902,27 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
...
+	if err != nil {
+		return
+	}
+
+	serverVersion, err := client.Version(cmd.Context())
+	if err != nil {
+		return
+	}
+
+	if serverVersion != version.Version {
```
Comment (pdevine) [inline cmd/cmd.go:918]: I think it's great to only show the single version unless there's a mismatch, but I think this should probably be more english text. It should look like:
```
Ollama version: 0.1.2
Warning: Your client is using version 0.1.3
```

File: cmd/cmd.go [hunk 6]
```diff
@@ -98,7 +99,7 @@ func CreateHandler(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
-	if err := client.Create(context.Background(), &request, fn); err != nil {
+	if err := client.Create(cmd.Context(), &request, fn); err != nil {
```
Comment (pdevine) [inline cmd/cmd.go:102]: I thought the original reason why we didn't reuse the context from cobra was that it was weird to reuse that w/ the http context?

**Review by pdevine — COMMENTED** [review]
The version handler looks good. Super useful.

File: cmd/cmd.go (hunk 6 above)
Comment (mxyng) [reply cmd/cmd.go:102]: I haven't seen any issues with using cobra.Context

File: cmd/cmd.go (hunk 4 above)
Comment (mxyng) [reply cmd/cmd.go:906]: The client version is the main thing since the server may not be available

File: cmd/cmd.go (hunk 6 above)
Comment (pdevine) [reply cmd/cmd.go:102]: Seems fine to me, but it's slightly weird to reuse the context from cobra to pass it to the client.

File: cmd/cmd.go (hunk 4 above)
Comment (pdevine) [reply cmd/cmd.go:906]: Wouldn't we tell you that you can't connect to Ollama in that case? It seems like the version is somewhat beside the point.

Comment (mxyng) [reply cmd/cmd.go:906]: Those are bad examples and "client" is not the right term. It's the version of the `ollama` command.

If the server doesn't start and you want to know why, a first step would be to identify the version.

Similarly if some other subcommand isn't able to connect to the server, a first step would be to get the version and try to reproduce. Whether it's connectivity related or not is irrelevant.

The version of the command is valuable by itself and should be the main focus.

File: cmd/cmd.go (hunk 6 above)
Comment (mxyng) [reply cmd/cmd.go:102]: It's not necessarily context from cobra.Command. cobra.Command is being passed in a context which can be retrieved with cmd.Context(). See https://github.com/jmorganca/ollama/blob/main/main.go#L11.

This allow the same context to be used by multiple command, specifically client requests, which is useful for shared data

File: cmd/cmd.go (hunk 3 above)
Comment (mxyng) [reply cmd/cmd.go:861]: On top of reducing duplication, this is less error prone than requiring each subcommand to register their own PreRun. Only a small subset of commands should _not_ check server heartbeat. This changes the check to an opt-out operation rather than an opt-in operation which is less error prone.

File: cmd/cmd.go (hunk 4 above)
Comment (pdevine) [reply cmd/cmd.go:830]: I disagree here. The version of the server running the LLM is the key because it can lead to issues not being able to run the model correctly. That's typically where people are going to run into problems. The client (command?) mismatch is also a problem, but that's accounted for by showing the warning.

If we also want to show the client version for cases where you can't connect, I think I'm OK with a similar string to "You client is using version ...", but again, it shouldn't be the main focus.

File: cmd/cmd.go (hunk 3 above)
Comment (pdevine) [reply cmd/cmd.go:861]: I'd rather be explicit here and than implicitly calling it. It's a single line per command.

File: cmd/cmd.go [hunk 7]
```diff
@@ -875,8 +868,27 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
...
+	fmt.Printf("ollama version %s\n", version.Version)
+
+	client, err := api.ClientFromEnvironment()
+	if err != nil {
+		return
+	}
+
+	serverVersion, err := client.Version(cmd.Context())
+	if err != nil {
+		return
```
Comment (BruceMacD) [inline cmd/cmd.go:881]: I think using the client version and not logging when the server version check fails could lead to a scenario where we are troubleshooting on the wrong version.

Ex: The server fails to start or crashes, and when a user reports the error they send the version from the CLI, which may not be accurate due to the server not being responsive.

**Review by skydiver1980 — APPROVED** [review]

File: cmd/cmd.go [hunk 8]
```diff
@@ -549,19 +542,35 @@ func generate(cmd *cobra.Command, opts generateOptions) error {
...
 		return err
 	}
+
 	if opts.Prompt != "" {
 		fmt.Println()
 		fmt.Println()
 	}
 
-	if !latest.Done {
+	if canceled {
```
Comment (BruceMacD) [inline cmd/cmd.go:569]: was this a bug? seems unrelated to this change

File: cmd/cmd.go [hunk 9]
```diff
@@ -1028,8 +1034,29 @@ func checkServerHeartbeat(_ *cobra.Command, _ []string) error {
...
+		return
+	}
+
+	serverVersion, err := client.Version(cmd.Context())
+	if err != nil {
+		fmt.Println("Warning: Your server is not accessible")
+	}
+
+	if serverVersion != "" {
+		fmt.Printf("Your ollama version %s\n", serverVersion)
```
Comment (BruceMacD) [inline cmd/cmd.go:1049]:
```suggestion
		fmt.Printf("%s\n", serverVersion)
```

**Review by BruceMacD — APPROVED** [review]

**Review by BruceMacD — APPROVED** [review]

**Review by pdevine — APPROVED** [review]
LGTM.
